### ISSUE TITLE
Insane memory usage (7GB+)

### ISSUE THREAD

**Issue by CherryDT** [stateful/runme#99]
Type: <b>Bug</b>

I noticed that there is a `node` process in WSL1 which is eating insane amounts of memory (7 GB) and occupies one CPU core fully. This seems to be the extension host. Extension Bisect showed that `stateful.runme` is the culprit. Disabling this extension resolves the issue.

`top` output:

![image](https://user-images.githubusercontent.com/6010463/211687858-c1d098dd-af1a-48b6-a03e-201f0c1da616.png)

Performance graph in Process Hacker:

![image](https://user-images.githubusercontent.com/6010463/211687885-3f3b8813-0dff-4c62-8133-0200ca14e0dc.png)

VS Code version: Code 1.74.3 (97dec172d3256f8ca4bfb2143f3f76b503ca0534, 2023-01-09T16:59:02.252Z)
OS version: Windows_NT x64 10.0.19045
Modes:
Sandboxed: No
Remote OS version: Linux x64 4.4.0-19041-Microsoft
Remote OS version: Linux x64 4.4.0-19041-Microsoft

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 PRO 4750U with Radeon Graphics (16 x 1697)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|undefined|
|Memory (System)|31.23GB (0.36GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu-20.04|
|OS|Linux x64 4.4.0-19041-Microsoft|
|CPUs|AMD Ryzen 7 PRO 4750U with Radeon Graphics (16 x 1700)|
|Memory (System)|31.23GB (0.34GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu-20.04|
|OS|Linux x64 4.4.0-19041-Microsoft|
|CPUs|AMD Ryzen 7 PRO 4750U with Radeon Graphics (16 x 1700)|
|Memory (System)|31.23GB (0.34GB free)|
|VM|0%|
</details><details><summary>Extensions (64)</summary>

Extension|Author (truncated)|Version
---|---|---
auto-close-tag|for|0.5.14
vscode-graphql-syntax|Gra|1.0.5
dotenv|mik|1.0.1
vscode-aql|mon|1.7.0
jupyter-keymap|ms-|1.0.0
remote-containers|ms-|0.266.1
remote-ssh|ms-|0.94.0
remote-ssh-edit|ms-|0.84.0
remote-wsl|ms-|0.72.0
remote-explorer|ms-|0.0.3
ejs-language-support|Qas|0.0.1
vscode-todo-highlight|way|1.0.5
vscode-devdocs|akf|1.0.3
svelte-intellisense|ard|0.7.1
vscode-intelephense-client|bme|1.9.3
npm-intellisense|chr|1.4.4
js-codeformer|cms|2.6.1
compulim-vscode-closetag|Com|1.2.0
vscode-eslint|dba|2.2.6
devdocs|dei|0.2.0
vscode-new-file|dku|4.0.2
dbux-code|Dom|0.7.9
xml|Dot|2.5.1
EditorConfig|Edi|0.16.4
copilot|Git|1.65.7705
vscode-graphql-execution|Gra|0.1.6
vscode-graphql-syntax|Gra|1.0.5
vscode-git-blamer|how|1.1.2
join-comment-aware|joh|0.0.3
solidity|Jua|0.0.141
edge|luo|0.3.2
rainbow-csv|mec|3.5.0
csharp|ms-|1.25.2
vscode-dotnet-runtime|ms-|1.6.0
isort|ms-|2022.8.0
python|ms-|2022.20.1
vscode-pylance|ms-|2023.1.10
jupyter|ms-|2022.11.1003412109
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.12
vscode-jupyter-cell-tags|ms-|0.1.6
vscode-jupyter-slideshow|ms-|0.1.5
cpptools|ms-|1.13.9
hexeditor|ms-|1.9.9
hexeditor|not|1.8.2
vscode-print|pdc|0.10.20
svelte-extractor|pro|0.0.3
vscode-data-preview|Ran|2.3.0
vscode-data-table|Ran|1.12.0
vscode-yaml|red|1.11.0
ActiveFileInStatusBar|Ros|1.0.3
vscode-paste-and-indent|Rub|0.0.8
bracket-jumper|sas|1.1.8
trailing-spaces|sha|0.4.1
vscode-standard|sta|2.1.3
runme|sta|0.4.2
ignore-gitignore|stu|1.0.1
svelte-vscode|sve|107.0.1
es6-string-html|Tob|2.12.0
sort-lines|Tyr|1.9.1
use-strict-everywhere|vee|0.1.3
bracket-padder|via|0.3.0
change-case|wma|1.0.0
html-css-class-completion|Zig|1.20.0

</details>

**Comment by admc** [comment]
Thank you very much for the bug report @CherryDT !

**Comment by sourishkrout** [comment]
The bug report's much appreciated @CherryDT. We'll get to the bottom of this mem consumption issue.

**Comment by mxsdev** [comment]
I was able to reproduce this on WSL1 with Windows 10. 

After some investigation, it looks like there's a memory leak in [yuin/goldmark][ast]. @CherryDT if you wouldn't mind checking out [this demo repo I made][demo] and running the test script (`node ./wasm/test.js`) in WSL, that would be great to see if this is the source of the issue on your machine.

Even on my macos machine, the program eats up ~3.5 GB before exiting (see below). That probably shouldn't happen in the first place, but I think the memory does get freed on other platforms. I'm guessing that on certain platforms, that memory is getting leaked.

If you have the same issue as below, I'll leave an issue on the goldmark repo, since I couldn't find any reports there already.

![unnamed](https://user-images.githubusercontent.com/16108792/213218249-7f64c6ea-4cf6-4610-a66b-621c493f85ca.gif)

[ast]: https://github.com/yuin/goldmark
[demo]: https://github.com/mxsdev/runme-goldmark-perf

**Comment by adambabik** [comment]
@mxsdev thanks for digging into this. It's interesting case because in your test, in the Go code, you don't actually do any work except for declaring a global variable. However, the initial bump in the memory usage is quite odd.

On macOS, I managed to reproduce it, but after a while the used memory drops to around 300-400 MB. I observed that with these two changes:

```js
// wasm/test.js
;(async function main() {
    await initWasm()

    await new Promise(resolve => setTimeout(resolve, 60000));
})()
```

```go
func main() {
    select {}
}
```

I also run a test where, instead of just declaring a global variable, it performs some work:

```go
package main

import (
        "github.com/yuin/goldmark"
        "github.com/yuin/goldmark/text"
)

func main() {
        md := []byte("## markdown document")
        p := goldmark.DefaultParser()
        _ = p.Parse(text.NewReader(md))

        select {}
}
```

Interestingly, there is no spike of the used mem to 3.5GB but to around 800MB and it drops quickly to less than 400MB. It might be also an issue of how I observe memory usage generally and the spike is there, but it lasts much shorter.

I will try to benchmark our wasm binary with the whole logic, but I would probably blame Go and its WebAssmebly implementation and/or how we use it rather than goldmark at this point.

**Comment by mxsdev** [comment]
@adambabik Ah, I see - that definitely suggests it may be an issue at the level of Go's wasm compilation/how we use it, as you say.

Definitely interested to see the benchmarks. I'll keep looking into it on my own as well 👍

**Comment by sourishkrout** [comment]
Thanks for the extra contextual information folks 👏 . If it's possible to patch up the memory leak... let's do it!

Outside of that: Based on some digging we've done a few weeks back, we believe the culprit here is likely how we're using the WASM Engine inside of the the Extension Host (node process). The VS Code team has pointed me in the direction to wrap the WASM module inside of a worker thread (message passing) instead of sharing the memory/process space.

However, doing exactly that is basically akin to a larger shift in architecture which is currently under way to give Runme a RPC interface. To get a better read on whether it will solve/mitigate this issue we decided to expedite moving De(-)serialization to this interface (currently provided by WASM integration) shorter term. This is actively being worked on and hopefully wraps in the next few weeks.

Long story short is that the WASM integration will go away. I suppose a memor leak might still be problematic in the "runme server".

**Comment by CherryDT** [comment]
> After some investigation, it looks like there's a memory leak in [yuin/goldmark](https://github.com/yuin/goldmark). @CherryDT if you wouldn't mind checking out [this demo repo I made](https://github.com/mxsdev/runme-goldmark-perf) and running the test script (`node ./wasm/test.js`) in WSL, that would be great to see if this is the source of the issue on your machine.

Yes this seems to be it.

Again one CPU core fully utilized and 8 GB RAM usage.

![image](https://user-images.githubusercontent.com/6010463/213291681-8eb3bb89-9dee-4c5f-b932-2b8b20c7b429.png)

**Comment by sfxworks** [comment]
I believe that I am seeing this too, but since I run code-server with 4GB of ram, it just crashes over and over. 

```sh
[07:18:42] [10.0.3.238][ebf03487][ExtensionHostConnection] <2428> Extension Host Process exited with code: null, signal: SIGKILL.
[07:18:42] [10.0.3.238][ebf03487][ExtensionHostConnection] Unknown reconnection token (seen before).
[07:18:43] [10.0.3.238][323fff64][ExtensionHostConnection] New connection established.
[07:18:43] [10.0.3.238][323fff64][ExtensionHostConnection] <2558> Launched Extension Host Process.
[07:18:57] [10.0.3.238][323fff64][ExtensionHostConnection] <2558> Extension Host Process exited with code: null, signal: SIGKILL.
[07:18:57] [10.0.3.238][323fff64][ExtensionHostConnection] Unknown reconnection token (seen before).
[07:18:58] [10.0.3.238][a5c229ce][ExtensionHostConnection] New connection established.
[07:18:58] [10.0.3.238][a5c229ce][ExtensionHostConnection] <2847> Launched Extension Host Process.
[07:19:10] [10.0.3.238][a5c229ce][ExtensionHostConnection] <2847> Extension Host Process exited with code: null, signal: SIGKILL.
[07:19:10] [10.0.3.238][a5c229ce][ExtensionHostConnection] Unknown reconnection token (seen before).
[07:20:40] [10.0.3.238][16488969][ExtensionHostConnection] New connection established.
[07:20:40] [10.0.3.238][16488969][ExtensionHostConnection] <2979> Launched Extension Host Process.
[07:20:53] [10.0.3.238][16488969][ExtensionHostConnection] <2979> Extension Host Process exited with code: null, signal: SIGKILL.
[07:20:53] [10.0.3.238][16488969][ExtensionHostConnection] Unknown reconnection token (seen before).
```

**Comment by sourishkrout** [comment]
Hi @sfxworks! Thanks for letting us know. We're working to resolve this mem issue as part of moving Runme's architecture away from its current WASM lib implementation. Planning to have a fix ready before the end of the month! Thanks for giving Runme a spin. This bug is on top of our priority list.

**Comment by sourishkrout** [comment]
Hello @CherryDT & @sfxworks 👋 ! We just released the Runme extension (https://marketplace.visualstudio.com/items?itemName=stateful.runme) version 0.7.0, which transitions away from the WebAssembly-based parsing that's led to the crashing extension & extension host. We'd love for you to try it, and let us know what you think.

Please note that if you're on Windows, Runme has yet to support PowerShell fully (pending). However, it should no longer lead to any crashes. We've tested it thoroughly. 

Please feel free to let us know if you're running into any issues. Here or on Discord: https://discord.gg/stateful. Enjoy your weekend 😄!

**Comment by sfxworks** [comment]
Perfect. Thank you for the tag! We use code-server on our end so no worries about the Windows of things.

**Comment by sourishkrout** [comment]
> Perfect. Thank you for the tag! We use code-server on our end so no worries about the Windows of things.

Great. Let us know how it goes!

**Comment by sourishkrout** [comment]
Closing this issue as fixed, @CherryDT. Let me know if you're still seeing issues.
